### PR TITLE
Add support for policy_update_patterns in  policy_set

### PR DESCRIPTION
**Pull request by nithishravindra**
## Description

Add PolicyUpdatePatterns support for policy sets (read/create/update) and tests

_Remember to:_

- [X] _Update the [Change Log](https://github.com/hashicorp/terraform-provider-tfe/blob/main/docs/changelog-process.md)_
- [X] _Update the [Documentation](https://github.com/hashicorp/terraform-provider-tfe/blob/main/docs/changelog-process.md#updating-the-documentation)_

## Testing plan

1.  _Describe how to replicate_
1.  _the conditions_
1.  _under which your code performs its purpose,_
1.  _including example Terraform configs where necessary._

## External links

_Include any links here that might be helpful for people reviewing your PR. If there are none, feel free to delete this section._

- [go-tfe documentation](https://pkg.go.dev/github.com/hashicorp/go-tfe?tab=doc#xxxx)
- [go-tfe-PR](https://github.com/hashicorp/go-tfe/pull/1306)
- [API documentation](https://developer.hashicorp.com/terraform/cloud-docs/api-docs/policy-sets#create-a-policy-set)
- [Related PR](https://github.com/hashicorp/terraform-provider-tfe/pull/xxxx)

## Output from acceptance tests

_Please run applicable acceptance tests locally and include the output here. See [testing.md](https://github.com/hashicorp/terraform-provider-tfe/blob/main/docs/testing.md) to learn how to run acceptance tests._

_If you are an external contributor, your contribution(s) will first be reviewed before running them against the project's CI pipeline._

```
resource "tfe_policy_set" "dummy_patterns" {
  name                = "dummy-policy-set-update-patterns"
  description         = "Dummy policy set for update pattern testing"
  organization        = data.tfe_organization.test.name
  kind                = "sentinel"
  agent_enabled       = true
  policy_tool_version = "0.24.1"
  workspace_ids       = ["ws-x"]
  policy_update_patterns = [
    "policies/dummy-policy-set/*",
    "sentinel/*.sentinel",
  ]

  vcs_repo {
    identifier                 = "xxx/xxx"
    ingress_submodules         = false
    github_app_installation_id = "ghain-xxx"
  }
}

output "dummy_policy_set_id" {
  value = tfe_policy_set.dummy_patterns.id
}

...
```

### CODE REVIEW

File: internal/provider/resource_tfe_policy_set.go [hunk 1]
```diff
@@ -408,6 +422,12 @@ func resourceTFEPolicySetUpdate(d *schema.ResourceData, meta interface{}) error
 			options.PolicyToolVersion = tfe.String(policyToolVersion.(string))
 		}
 
+		if d.HasChange("policy_update_patterns") {
+			for _, pattern := range d.Get("policy_update_patterns").([]interface{}) {
```
Comment (sana-faraz) [inline internal/provider/resource_tfe_policy_set.go:426]: nit: iteration can be moved to another function and reused

File: internal/provider/resource_tfe_policy_set.go [hunk 2]
```diff
@@ -383,7 +397,7 @@ func resourceTFEPolicySetUpdate(d *schema.ResourceData, meta interface{}) error
 	}
 
 	// Don't bother updating the policy set's attributes if they haven't changed
-	if d.HasChange("name") || d.HasChange("description") || d.HasChange("global") || d.HasChange("vcs_repo") || d.HasChange("overridable") || d.HasChange("agent_enabled") || d.HasChange("policy_tool_version") {
+	if d.HasChange("name") || d.HasChange("description") || d.HasChange("global") || d.HasChange("vcs_repo") || d.HasChange("overridable") || d.HasChange("agent_enabled") || d.HasChange("policy_tool_version") || d.HasChange("policy_update_patterns") {
```
Comment (sana-faraz) [inline internal/provider/resource_tfe_policy_set.go:400]: nit:  Can be rewitten like
```
fields := []string{
    "name", "description", "global", "vcs_repo", 
    "overridable", "agent_enabled", "policy_tool_version", 
    "policy_update_patterns",
}

hasAnyChange := false
for _, field := range fields {
    if d.HasChange(field) {
        hasAnyChange = true
        break
    }
}

if hasAnyChange {
    ....
}
```

File: internal/provider/data_source_policy_set_test.go [hunk 3]
```diff
@@ -401,6 +407,7 @@ resource "tfe_policy_set" "foobar" {
   name         = "tst-policy-set-%d"
   description  = "Policy Set"
   organization = local.organization_name
+	policy_update_patterns = ["**/*.sentinel", "policies/**/*.hcl"]
```
Comment (shrishyamb) [inline internal/provider/data_source_policy_set_test.go]: nit
```suggestion
  policy_update_patterns = ["**/*.sentinel", "policies/**/*.hcl"]
```

File: internal/provider/resource_tfe_policy_set_test.go [hunk 4]
```diff
@@ -1258,6 +1282,7 @@ resource "tfe_policy_set" "foobar" {
   name         = "tst-terraform"
   description  = "Policy Set"
   organization     = local.organization_name
+	policy_update_patterns = ["**/*.sentinel", "policies/**/*.hcl"]
```
Comment (shrishyamb) [inline internal/provider/resource_tfe_policy_set_test.go]: nit
```suggestion
  policy_update_patterns = ["**/*.sentinel", "policies/**/*.hcl"]
```

File: internal/provider/resource_tfe_policy_set_test.go [hunk 5]
```diff
@@ -1284,6 +1309,7 @@ resource "tfe_policy_set" "foobar" {
...
```
Comment (shrishyamb) [inline internal/provider/resource_tfe_policy_set_test.go]: nit
```suggestion
  policy_update_patterns = ["**/*.sentinel", "policies/**/*.hcl"]
```

File: internal/provider/resource_tfe_policy_set_test.go [hunk 6]
```diff
@@ -1318,6 +1344,7 @@ resource "tfe_policy_set" "foobar" {
...
```
Comment (shrishyamb) [inline internal/provider/resource_tfe_policy_set_test.go]: nit
```suggestion
  policy_update_patterns = ["**/*.sentinel", "policies/**/*.hcl"]
```

File: go.mod [hunk 7]
```diff
@@ -12,7 +12,7 @@ require (
 	github.com/hashicorp/go-multierror v1.1.1 // indirect
 	github.com/hashicorp/go-retryablehttp v0.7.8 // indirect
 	github.com/hashicorp/go-slug v1.0.0
-	github.com/hashicorp/go-tfe v1.103.0
+	github.com/hashicorp/go-tfe v1.103.1-0.20260408065133-6253066cb49c
```
Comment (shrishyamb) [inline go.mod]: I see the go-tfe is released officially with [v1.104.0](https://github.com/hashicorp/go-tfe/releases/tag/v1.104.0). Consider updating the version to use official one.

File: internal/provider/resource_tfe_policy_set.go [hunk 8]
```diff
@@ -209,6 +216,12 @@ func resourceTFEPolicySetCreate(d *schema.ResourceData, meta interface{}) error
 		options.PolicyToolVersion = tfe.String(vPolicyToolVersion.(string))
 	}
 
+	if vPolicyUpdatePatterns, ok := d.GetOk("policy_update_patterns"); ok {
+		for _, pattern := range vPolicyUpdatePatterns.([]interface{}) {
```
Comment (shrishyamb) [inline internal/provider/resource_tfe_policy_set.go:220]: Missing the empty-list case. If the user sets policy_update_patterns = [], the loop produces nothing and options.PolicyUpdatePatterns stays nil. Refer to how `trigger_patterns` handles this in `resource_tfe_workspace.go`. Also consider adding missing a test for updating policy_update_patterns from a non-empty list to [].

**Review by shrishyamb — COMMENTED** [review]
Added comments. Also consider updating the PR description with test plan and acceptance test output as per the guide.

File: internal/provider/data_source_policy_set_test.go (hunk 3 above)
Comment (sana-faraz) [reply internal/provider/data_source_policy_set_test.go]: Done

File: internal/provider/resource_tfe_policy_set.go (hunk 8 above)
Comment (sana-faraz) [reply internal/provider/resource_tfe_policy_set.go:220]: Done

File: internal/provider/resource_tfe_policy_set.go (hunk 2 above)
Comment (sana-faraz) [reply internal/provider/resource_tfe_policy_set.go]: Done

File: internal/provider/resource_tfe_policy_set.go (hunk 1 above)
Comment (sana-faraz) [reply internal/provider/resource_tfe_policy_set.go]: Done

File: internal/provider/resource_tfe_policy_set_test.go (hunk 4 above)
Comment (sana-faraz) [reply internal/provider/resource_tfe_policy_set_test.go]: Done

File: internal/provider/resource_tfe_policy_set_test.go (hunk 5 above)
Comment (sana-faraz) [reply internal/provider/resource_tfe_policy_set_test.go]: Done

File: internal/provider/resource_tfe_policy_set_test.go (hunk 6 above)
Comment (sana-faraz) [reply internal/provider/resource_tfe_policy_set_test.go]: Done

File: go.mod (hunk 7 above)
Comment (sana-faraz) [reply go.mod]: Rebased

File: CHANGELOG.md [hunk 9]
```diff
@@ -1,11 +1,13 @@
 ## Unreleased
 ENHANCEMENTS:
 * Updates warning when using credentials/config file for authentication and running on cloud to be clearer, by @christian-doucette [#2036](https://github.com/hashicorp/terraform-provider-tfe/pull/2036)
+* Support trigger patterns and working directories in stacks by @aaabdelgany [#2048](https://github.com/hashicorp/terraform-provider-tfe/pull/2048)
 
 ## v0.76.2
 
 FEATURES:
 * `r/tfe_notification_configuration`: Optionally set `url` as a write-only (`url_wo`). Adds support for auto-managed (via hash) for `token_wo` param to avoid manually managing `_version` arguments. By @drewmullen [#2028](https://github.com/hashicorp/terraform-provider-tfe/pull/2028)
+- `r/tfe_policy_set` and `d/tfe_policy_set`: Add `policy_update_patterns` attribute support for policy sets. By @nithishravindra [#2030](https://github.com/hashicorp/terraform-provider-tfe/pull/2030)
```
Comment (ctrombley) [inline CHANGELOG.md]: This entry should be under the Unreleased section.

File: CHANGELOG.md [hunk 10]
```diff
@@ -1,11 +1,13 @@
 ## Unreleased
 ENHANCEMENTS:
 * Updates warning when using credentials/config file for authentication and running on cloud to be clearer, by @christian-doucette [#2036](https://github.com/hashicorp/terraform-provider-tfe/pull/2036)
+* Support trigger patterns and working directories in stacks by @aaabdelgany [#2048](https://github.com/hashicorp/terraform-provider-tfe/pull/2048)
```
Comment (ctrombley) [inline CHANGELOG.md:5]: Is this intentional? It's about a different PR.

File: internal/provider/resource_tfe_policy_set.go (hunk 8 above)
Comment (ctrombley) [reply internal/provider/resource_tfe_policy_set.go:220]: I might be missing something, but I don't see the requested test in this PR.

**Review by ctrombley — COMMENTED** [review]
This is looking good, thanks @nithishravindra. Just a few comments here.

File: CHANGELOG.md (hunk 9 above)
Comment (sana-faraz) [reply CHANGELOG.md]: Done

File: CHANGELOG.md (hunk 10 above)
Comment (sana-faraz) [reply CHANGELOG.md:5]: Maybe an issue during merge conflict resolution, rebased

**Review by ctrombley — APPROVED** [review]
Looks good, thanks @nithishravindra!
